packages/core/src/tracing/instrumentation/cloud/aws/s3.js

```diff
@@ -0,0 +1,215 @@
+/*
+ * (c) Copyright IBM Corp. 2021
+ * (c) Copyright Instana Inc. and contributors 2021
+ */
+
+'use strict';
+
+const shimmer = require('shimmer');
+const cls = require('../../../cls');
+const { EXIT, isExitSpan } = require('../../../constants');
+const requireHook = require('../../../../util/requireHook');
+const tracingUtil = require('../../../tracingUtil');
+
+const s3MethodsInfo = {
+  listBuckets: {
+    op: 'list',
+    hasKey: false
+  },
+  createBucket: {
+    op: 'createBucket',
+    hasKey: false
+  },
+  deleteBucket: {
+    op: 'deleteBucket',
+    hasKey: false
+  },
+  headObject: {
+    op: 'metadata',
+    hasKey: true
+  },
+  putObject: {
+    op: 'put',
+    hasKey: true
+  },
+  deleteObject: {
+    op: 'delete',
+    hasKey: true
+  },
+  getObject: {
+    op: 'get',
+    hasKey: true
+  },
+  listObjects: {
+    op: 'list',
+    hasKey: false
+  },
+  listObjectsV2: {
+    op: 'list',
+    hasKey: false
+  }
+};
+
+const methodList = Object.keys(s3MethodsInfo);
+
+let isActive = false;
+
+exports.init = function init() {
+  requireHook.onModuleLoad('aws-sdk', instrumentAWS);
+};
+
+exports.activate = function activate() {
+  isActive = true;
+};
+
+exports.deactivate = function deactivate() {
+  isActive = false;
+};
+
+function instrumentAWS(AWS) {
+  shimmer.wrap(AWS.Service.prototype, 'makeRequest', shimMakeRequest);
+  shimmer.wrap(AWS.Request.prototype, 'promise', shimRequestPromise);
+}
+
+function shimMakeRequest(originalMakeRequest) {
+  return function() {
+    if (isActive) {
+      const originalArgs = new Array(arguments.length);
+      for (let i = 0; i < originalArgs.length; i++) {
+        originalArgs[i] = arguments[i];
+      }
+
+      return instrumentedMakeRequest(this, originalMakeRequest, originalArgs);
+    }
+
+    return originalMakeRequest.apply(this, arguments);
+  };
+}
+
+function instrumentedMakeRequest(ctx, originalMakeRequest, originalArgs) {
+  /**
+   * Original args for ALL AWS SDK Requets: [method, params, callback]
+   */
+
+  if (!isOperationShimmable(originalArgs[0]) || typeof originalArgs[2] !== 'function') {
+    return originalMakeRequest.apply(ctx, originalArgs);
+  }
+
+  if (cls.tracingSuppressed()) {
+    return originalMakeRequest.apply(ctx, originalArgs);
+  }
+
+  const parentSpan = cls.getCurrentSpan();
+
+  if (!parentSpan || isExitSpan(parentSpan)) {
+    return originalMakeRequest.apply(ctx, originalArgs);
+  }
+
+  return cls.ns.runAndReturn(() => {
+    const span = cls.startSpan('s3', EXIT);
+    span.ts = Date.now();
+    span.stack = tracingUtil.getStackTrace(instrumentedMakeRequest);
+    span.data.s3 = buildS3Data(originalArgs[0], originalArgs[1]);
+
+    const _originalCallback = originalArgs[2];
+
+    originalArgs[2] = cls.ns.bind(function(err, data) {
+      finishSpan(err, data, span);
+      setImmediate(() => finishSpan(err, data, span));
```

Comment:
Pretty sure we don't need this with setImmediate here. The synchronous finishSpan in line 117 should be enough. This was only relevant for the SQS entry.